File: typescript/08_strings.ts
// ============================================================
// Topic:  Strings  (FOCUS: methods & operations)
// Summary: length, indexing, slicing, case, search, split/join, format.
// Run:     npx ts-node typescript/08_strings.ts
//
// KEY DIFFERENCE the user cares about:
//   TS: s.length, s.toLowerCase()   |   Python: len(s), s.lower()
// ============================================================

const s = "Hello, World";

// ----- 1. Length (a PROPERTY, no parentheses; Python: len(s)) -----
console.log(s.length); // 12   ⚠ s.length NOT len(s)

// ----- 2. Indexing (no negative indexing like Python) -----
console.log(s[0]); // H     | Python: s[0]
console.log(s[s.length - 1]); // d     | Python: s[-1]   (TS has no s[-1])
console.log(s.charAt(0)); // H     (alternative)
console.log(s.at(-1)); // d     (.at() DOES allow negatives)

// ----- 3. Slicing -> s.slice(start, stop)  (Python: s[start:stop]) -----
console.log(s.slice(0, 5)); // Hello
console.log(s.slice(7)); // World
console.log(s.split("").reverse().join("")); // dlroW ,olleH (reverse; Python: s[::-1])

// ----- 4. Case methods -----
console.log(s.toLowerCase()); // hello, world   | Python: s.lower()
console.log(s.toUpperCase()); // HELLO, WORLD   | Python: s.upper()

// ----- 5. Search / test -----
console.log(s.indexOf("World")); // 7   (-1 if missing) | Python: s.find("World")
console.log(s.includes("World")); // true               | Python: "World" in s
console.log(s.startsWith("Hello")); // true             | Python: s.startswith(...)
console.log(s.endsWith("World")); // true               | Python: s.endswith(...)

// ----- 6. Modify (strings are IMMUTABLE; methods return NEW strings) -----
console.log(s.replace("World", "Vetri")); // Hello, Vetri | Python: s.replace(...)
console.log(s.replaceAll("l", "L")); // HeLLo, WorLd
console.log("  pad  ".trim()); // 'pad'             | Python: s.strip()
console.log("a,b,c".split(",")); // [ 'a', 'b', 'c' ] | Python: "a,b,c".split(",")
console.log(["a", "b", "c"].join("-")); // a-b-c        | Python: "-".join([...])
console.log("hi".repeat(3)); // hihihi             | Python: "hi" * 3
console.log("5".padStart(3, "0")); // 005           | Python: "5".zfill(3)

// ----- 7. Formatting (template literals; Python: f-strings) -----
const name = "Vetri",
  age = 22;
console.log(`${name} is ${age}`); // Vetri is 22
console.log((3.14159).toFixed(2)); // 3.14  | Python: f"{x:.2f}"

// ----- 8. Char <-> code -----
console.log("A".charCodeAt(0)); // 65   | Python: ord("A")
console.log(String.fromCharCode(66)); // B    | Python: chr(66)

// ----- Quick reference table (TS -> Python) -----
// s.length          -> len(s)
// s.toLowerCase()   -> s.lower()
// s.toUpperCase()   -> s.upper()
// s.indexOf(x)      -> s.find(x)
// s.includes(x)     -> x in s
// s.trim()          -> s.strip()
// s.replaceAll(a,b) -> s.replace(a,b)
// s.split(",")      -> s.split(",")
// arr.join("-")     -> "-".join(arr)
// s.slice(a,b)      -> s[a:b]

File: rust/08_strings.rs
// ============================================================
// Topic:  Strings  (FOCUS: methods & operations)
// Summary: &str vs String, length, case, search, split/join, format.
// Run:     rustc rust/08_strings.rs -o rust/08_strings && rust/08_strings
//
// KEY DIFFERENCE the user cares about:
//   Rust: s.len(), s.to_lowercase()   |   Python: len(s), s.lower()
//   Rust: &str (slice) vs String (owned)
// ============================================================

fn main() {
    let s: &str = "Hello, World";

    // ----- 1. &str vs String -----
    let slice: &str = "hello"; // string slice — borrowed, fixed-size view
    let owned: String = String::from("hello"); // owned, growable
    let also_owned = "hello".to_string(); // another way to create String
    println!("{} {} {}", slice, owned, also_owned); // Output: hello hello hello

    // ----- 2. Length -----
    println!("{}", s.len()); // 12 (byte length)   | Python: len(s)   TS: s.length
    println!("{}", s.chars().count()); // 12 (Unicode scalar count)

    // ----- 3. Indexing (no s[0] — use chars or bytes) -----
    println!("{}", s.chars().next().unwrap()); // H   | Python: s[0]
    println!("{}", s.chars().last().unwrap()); // d   | Python: s[-1]

    // ----- 4. Slicing (byte indices — must be on char boundaries) -----
    println!("{}", &s[0..5]); // Hello
    println!("{}", &s[7..]); // World

    // ----- 5. Case methods -----
    println!("{}", s.to_lowercase()); // hello, world   | Python: s.lower()   TS: toLowerCase()
    println!("{}", s.to_uppercase()); // HELLO, WORLD   | Python: s.upper()   TS: toUpperCase()

    // ----- 6. Search / test -----
    println!("{}", s.find("World").unwrap()); // 7   | Python: s.find()   TS: indexOf()
    println!("{}", s.contains("World")); // true        | Python: "World" in s   TS: includes()
    println!("{}", s.starts_with("Hello")); // true     | TS: startsWith()
    println!("{}", s.ends_with("World")); // true       | TS: endsWith()
    println!("{}", s.matches('l').count()); // 3 (count occurrences)

    // ----- 7. Modify (strings are IMMUTABLE; methods return NEW strings) -----
    println!("{}", s.replace("World", "Vetri")); // Hello, Vetri
    println!("{}", "  pad  ".trim()); // pad         | TS: trim()
    println!("{:?}", "a,b,c".split(',').collect::<Vec<_>>()); // ["a", "b", "c"]
    println!("{}", ["a", "b", "c"].join("-")); // a-b-c   | Python: "-".join(list)
    println!("{}", "ha".repeat(3)); // hahaha         | Python: "ha" * 3

    // ----- 8. Formatting -----
    let name = "Vetri";
    let age = 22;
    println!("{} is {}", name, age); // Vetri is 22   | Python: f"{name} is {age}"
    println!("{}", format!("{} is {}", name, age)); // Vetri is 22 (builds String)

    // ----- 9. Char <-> code -----
    println!("{}", 'A' as u32); // 65   | Python: ord("A")   TS: charCodeAt(0)
    println!("{}", char::from_u32(66).unwrap()); // B   | Python: chr(66)

    // ----- Quick reference table (Rust -> Python/TS) -----
    // s.len()              -> len(s) / s.length
    // s.to_lowercase()     -> s.lower() / s.toLowerCase()
    // s.to_uppercase()     -> s.upper() / s.toUpperCase()
    // s.find(x)            -> s.find(x) / s.indexOf(x)
    // s.contains(x)        -> x in s / s.includes(x)
    // s.trim()             -> s.strip() / s.trim()
    // s.replace(a, b)      -> s.replace(a, b) / s.replaceAll(a, b)
    // s.split(",")         -> s.split(",") / s.split(",")
    // ["a","b"].join("-")  -> "-".join(list) / list.join("-")
    // &s[a..b]             -> s[a:b] / s.slice(a, b)
}